subprojects/core-api/src/main/java/org/gradle/work/FileChange.java

```diff
@@ -38,6 +38,16 @@
      */
     ChangeType getChangeType();
 
+    /**
+     * The file type of the file.
+     *
+     * <p>
+     *     For {@link ChangeType#ADDED} and {@link ChangeType#MODIFIED}, the type of the file which was added/modified is reported.
+     *     For {@link ChangeType#REMOVED} the type of the file which was removed is reported.
```

Comment:
Would be so nice if we had Kotlin's `when` operator, so we could do things like this:

```kotlin
when (change) {
    is Added -> // change.type is the added file type
    is Modified -> // change.originalType is the original type, change.newType is the current type
    is Removed -> // change.originalType is the original type
}
```

Reply:
If this was using a visitor pattern we could have different `visit*()` methods to this effect...

Reply:
Java 12 has switch expressions: https://openjdk.java.net/jeps/325

Maybe we can switch to a visitor pattern later on.

Anything you want me to do about this comment now?